File: IskrClockiOS/IskrClockiOS/Services/CustomStationsManager.swift
//
//  CustomStationsManager.swift
//  IskrClockiOS
//

import Foundation

class CustomStationsManager: ObservableObject {
    @Published var customStations: [RadioStation] = []

    init() {
        loadStations()
    }

    func addStation(_ station: RadioStation) {
        customStations.append(station)
        saveStations()
    }

    func deleteStation(_ station: RadioStation) {
        customStations.removeAll { $0.id == station.id }
        saveStations()
    }

    func updateStation(_ station: RadioStation) {
        if let index = customStations.firstIndex(where: { $0.id == station.id }) {
            customStations[index] = station
            saveStations()
        }
    }

    private func loadStations() {
        guard let data = UserDefaults.standard.data(forKey: "customStations"),
              let stations = try? JSONDecoder().decode([RadioStation].self, from: data) else {
            customStations = []
            return
        }
        customStations = stations
    }

    private func saveStations() {
        if let encoded = try? JSONEncoder().encode(customStations) {
            UserDefaults.standard.set(encoded, forKey: "customStations")
        }
    }
}

File: IskrClockFlutter/lib/services/custom_stations_manager.dart
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:convert';
import '../models/radio_station.dart';

class CustomStationsManager with ChangeNotifier {
  List<RadioStation> _customStations = [];

  List<RadioStation> get customStations => _customStations;

  CustomStationsManager() {
    _loadStations();
  }

  Future<void> _loadStations() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final stationsJson = prefs.getString('customStations');
      if (stationsJson != null) {
        final List<dynamic> decoded = json.decode(stationsJson);
        _customStations = decoded.map((json) => RadioStation.fromJson(json)).toList();
        notifyListeners();
      }
    } catch (e) {
      print('Error loading custom stations: $e');
    }
  }

  Future<void> _saveStations() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final encoded = json.encode(_customStations.map((s) => s.toJson()).toList());
      await prefs.setString('customStations', encoded);
    } catch (e) {
      print('Error saving custom stations: $e');
    }
  }

  Future<void> addStation(RadioStation station) async {
    _customStations.add(station);
    await _saveStations();
    notifyListeners();
  }

  Future<void> deleteStation(RadioStation station) async {
    _customStations.removeWhere((s) => s.id == station.id);
    await _saveStations();
    notifyListeners();
  }

  Future<void> updateStation(RadioStation station) async {
    final index = _customStations.indexWhere((s) => s.id == station.id);
    if (index != -1) {
      _customStations[index] = station;
      await _saveStations();
      notifyListeners();
    }
  }
}
